Add new config for rewrites/redirects/headers/etc, deprecate routes (#1498)
zeit/docs: components/references-mdx/configuration/configuration/configuration.mdx

**Type**: `Boolean`.

**Default Value**: `false`.

When set to `true`, both the [source view](/docs/v2/platform/deployments/#source-view) and [logs view](/docs/v2/platform/deployments/#logs-view) will be publicly accessible (when the user is authenticated with ZEIT).

<Code lang="json">{`{
  "public": true
}`}</Code>

## cleanUrls

**Type**: `Boolean`.

**Default Value**: `false`.

> When set to `true`, all HTML files will have their `.html` extension removed. When visiting a page that ends with `.html`, a 308 response will redirect the client to the extensionless page.

Reviewer: I believe we have used `path` in other places around the documentation, so probably better to be consistent.
```suggestion
When set to `true`, all HTML files will have their `.html` extension removed. When accessing a path that ends with `.html`, a 308 response will redirect the client to the extension-less path.
```

> For example, a file named `about.html` will be served when visiting the `/about` page. Visiting `/about.html` will redirect to `/about`.

Reviewer: ```suggestion
For example, a file named `about.html` will be served when visiting the `/about` path. Visiting `/about.html` will redirect to `/about`.
```

<Code lang="json">{`{
  "cleanUrls": true
}`}</Code>

## trailingSlash

**Type**: `Boolean`.

**Default Value**: `null`.

> When set to `true`, visiting a route that does not end with a forward slash will respond with a 308 status code and redirect to the route with a trailing slash.

Reviewer: `route` could be confused with `routes`.
```suggestion
When set to `true`, visiting a path that does not end with a forward slash will respond with a 308 status code and redirect to the path with a trailing slash.
```

> When set to `false`, visiting a route that ends with a forward slash will respond with a 308 status code and redirect to the route without the trailing slash.

Reviewer: ```suggestion
When set to `false`, visiting a path that ends with a forward slash will respond with a 308 status code and redirect to the path without the trailing slash.
```

> For example, when `trailingSlash: false` the `/about/` page will redirect to `/about`.

Reviewer: ```suggestion
For example, when `trailingSlash: false` the `/about/` path will redirect to `/about`.
```

<Code lang="json">{`{
  "public": true
}`}</Code>

## cleanUrls

**Type**: `Boolean`.

When set to `true`, all `.html` files will have their suffix removed. When visiting a page that ends with `.html`, a 301 redirect will be returned pointing to the extensionless page.

For example, a file named `about.html` will be served when visiting the `/about` page. Visiting `/about.html` will redirect to `/about`.

<Code lang="json">{`{
  "cleanUrls": true
}`}</Code>

## trailingSlash

**Type**: `Boolean`.

> When set to `true`, all routes that do not end with a trailing slash `/` will redirect to the same route with a trailing slash.

Author: I'm debating if we should document `true` here since it seems like a really strange use case.

Author: I removed it 😄

When set to `false`, all routes that end with a trailing slash `/` will redirect to the same route without the trailing slash.

For example, when `trailingSlash` is `false` then visiting the `/about/` page will redirect to `/about`.

<Code lang="json">{`{
  "trailingSlash": false
}`}</Code>

## redirects

**Type:** `Array` of redirect `Object`.

**Valid values:** a list of redirect definitions.

**Redirect object definition:**

> - `source`: A pattern that matches each incoming pathname (excluding querystring).

Author: @dav-is I noticed that the `routes.src` documentation above says "excluding querystring" so I added it here.

Is this still true?

Reviewer 2: It's still true

<Code lang="json">{`{
  "public": true
}`}</Code>

## cleanUrls

**Type**: `Boolean`.

When set to `true`, all HTML files will have their `.html` extension removed. When visiting a page that ends with `.html`, a 301 response will redirect the client to the extensionless page.

For example, a file named `about.html` will be served when visiting the `/about` page. Visiting `/about.html` will redirect to `/about`.

<Code lang="json">{`{
  "cleanUrls": true
}`}</Code>

## trailingSlash

**Type**: `Boolean`.

When set to `false`, visiting a route that ends with a forward slash will respond with a 307 status code and redirect to the route without the trailing slash.

For example, visiting the `/about/` page will redirect to `/about`.

<Code lang="json">{`{
  "trailingSlash": false
}`}</Code>

## redirects

**Type:** `Array` of redirect `Object`.

**Valid values:** a list of redirect definitions.

**Redirect object definition:**

- `source`: A pattern that matches each incoming pathname (excluding querystring).

> - `destination`: A location destination defined as a relative pathname or full URL.

Reviewer 2: ```suggestion
- `destination`: A location destination defined as an absolute pathname or external URL.
```

- `statusCode`: An optional HTTP status code in the [301-308](https://developer.mozilla.org/en-US/docs/Web/HTTP/Redirections) range (default 307).

<P>
  This example configures custom redirects that map to static files,{' '}
  <Link href="/docs/v2/serverless-functions/introduction">
    Serverless Functions
  </Link>,{' '}
  and external URLs.
</P>
<Code lang="json">{`{
  "redirects": [
    { "source": "/me", "destination": "/profile.html" },
    { "source": "/user", "destination": "/api/user", "statusCode": 308 },
    { "source": "/view-source", "destination": "https://github.com/zeit/now" }
  ]
}`}</Code>

## rewrites

**Type:** `Array` of rewrite `Object`.

**Valid values:** a list of rewrite definitions.

**Rewrite object definition:**

- `source`: A pattern that matches each incoming pathname (excluding querystring).

> - `destination`: A relative pathname to an existing file or a full URL.

Reviewer 2: ```suggestion
- `destination`: An absolute pathname to an existing resource or an external URL.
```

<P>
  This example configures custom rewrites that map to static files,{' '}
  <Link href="/docs/v2/serverless-functions/introduction">
    Serverless Functions
  </Link>,{' '}
  automatic query string matching, and a wildcard proxy.
</P>
<Code lang="json">{`{
  "rewrites": [
    { "source": "/about", "destination": "/about-our-company.html" },
    { "source": "/user/:id", "destination": "/api/user?id=:id" },
    { "source": "/resize/:width/:height", "destination": "/api/sharp" },
    { "source": "/proxy/(.*)", "destination": "https://example.com" }
  ]
}`}</Code>

## headers

**Type:** `Array` of header `Object`.

**Valid values:** a list of header definitions.

**Header object definition:**

- `source`: A pattern that matches each incoming pathname (excluding querystring).
- `headers`: An array of key/value pairs representing each response header.

<P>
  This example configures custom response headers for static files,{' '}
  <Link href="/docs/v2/serverless-functions/introduction">
    Serverless Functions

>   </Link>{' '}

Reviewer 3: ```suggestion
  </Link>{' '},
```
